### PR TITLE
[SPARK-35595][TESTS] Support multiple loggers in testing method withLogAppender

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

A test case of AdaptiveQueryExecSuite becomes flaky since there are too many debug logs in RootLogger:
https://github.com/Yikun/spark/runs/2715222392?check_suite_focus=true
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139125/testReport/

To fix it,  I suggest supporting multiple loggers in the testing method withLogAppender. So that the LogAppender gets clean target log outputs.

### Why are the changes needed?

Fix a flaky test case. 
Also, reduce unnecessary memory cost in tests.

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

Unit test

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 1]
```diff
@@ -1605,7 +1605,7 @@ class AdaptiveQueryExecSuite
   }
 
   test("SPARK-33933: Materialize BroadcastQueryStage first in AQE") {
-    val testAppender = new LogAppender("aqe query stage materialization order test")
+    val testAppender = new LogAppender("aqe query stage materialization order test", 10000)
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1608]: cc Max @MaxGekk since this max is from Max 
(https://github.com/apache/spark/commit/88fc8dbc09c5d24ae89413ab1e1fbabdf1fd8028#diff-1c74a76903c7da8f8424992b46b2f99157609726bc580d60e2d0858ea11c2aecR197) :D

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1608]: Do you know what kind of debug logs we have, and why so much. Maybe reduce them?

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1608]: I try running this locally, the log size is 388.
```
loggerName.map(Logger.getLogger).getOrElse(Logger.getRootLogger)
```
The test is flaky since it is using `Logger.getRootLogger`.
Shall we increase the default limit of LogAppender instead? What was the reason for adding a limit in the LogAppender?

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1608]: > What was the reason for adding a limit in the LogAppender?

I added the limit to prevent OOM and memory pressure in tests. It was added when I troubleshooted random OOMs in tests.

Comment (zhongyu09) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1608]: I guess many debug logs are introducing after https://github.com/apache/spark/pull/31269 ?

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1608]: That only added one debug log when materializing a stage, and a query won't have many stages.

Comment (zhongyu09) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1608]: I mean other feature's debug logs can also be collected by this LogAppender

File: core/src/test/scala/org/apache/spark/SparkFunSuite.scala [hunk 2]
```diff
@@ -217,19 +217,27 @@ abstract class SparkFunSuite
...
-      logger.setLevel(level.get)
+    val loggers = if (loggerNames.nonEmpty) {
+      loggerNames.map(Logger.getLogger)
+    } else {
+      Seq(Logger.getRootLogger)
+    }
+    val restoreLevels = loggers.map(_.getLevel)
+    loggers.foreach { logger =>
+      logger.addAppender(appender)
+      if (level.isDefined) {
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/SparkFunSuite.scala:231]: This could be `level.foreach`  in both places, but doesn't matter

**Review by srowen — COMMENTED** [review]
I noticed AQE tests timing out after stalling for hours in another PR - could be related

**Comment by gengliangwang** [comment]
> I noticed AQE tests timing out after stalling for hours in another PR - could be related

Yes the logger appending can be slow

**Comment by gengliangwang** [comment]
@HyukjinKwon @cloud-fan @MaxGekk @zhongyu09 I use a better approach and update the PR. Please take a look if you are interested. 
We should encourage developers to use specific loggers during code review.

**Comment by cloud-fan** [comment]
This is a good improvement! Do we still need to set the size to `10000`?

**Comment by gengliangwang** [comment]
> This is a good improvement! Do we still need to set the size to 10000?

Do you mean the default 1000 limit? 
We can keep it so that if it is hit we should enhance the problematic test case.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 3]
```diff
@@ -1605,15 +1605,17 @@ class AdaptiveQueryExecSuite
...
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:1608]: I mean here. Shall we revert this change? Not a big deal though

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala]: Hmmm I thought I already reverted it...Updated now.

**Comment by gengliangwang** [comment]
Thanks for the reviews, merging to master
